editor/src/clj/editor/workspace.clj

```diff
@@ -246,29 +256,66 @@ ordinary paths."
+(defn- find-clojure-plugins [workspace]
+  (let [resources (filter is-plugin-clojure-file? (g/node-value workspace :resource-list))]
     resources))
 
 (defn- load-plugin! [workspace resource]
   ; TODO Handle Exceptions!
   (log/info :msg (str "Loading plugin" (resource/path resource)))
-  (let [plugin (load-string (slurp resource))]
-    (plugin workspace))
+  (let [plugin-fn (load-string (slurp resource))]
+    (plugin-fn workspace))
   (log/info :msg (str "Loaded plugin" (resource/path resource))))
 
 (defn load-editor-plugins! [workspace added]
-  (reset! dbg-workspace workspace)
   (let [added-resources (set (map resource/proj-path added))
-        plugin-resources (find-editor-plugins workspace)
+        plugin-resources (find-clojure-plugins workspace)
         plugin-resources (filter (fn [x] (contains? added-resources (resource/proj-path x))) plugin-resources)]
     (dorun (map (fn [x] (load-plugin! workspace x)) plugin-resources))))
 
+; Determine if the extension has plugins, if so, it needs to be extracted
+
+(defn- is-jar-file? [resource]
+  (contains? #{"jar"} (resource/ext resource)))
+
+(defn- is-plugin-file? [resource]
+  (string/includes? (resource/proj-path resource) "/plugins/"))
+
+(defn- find-plugins-shared-libraries [workspace]
+  (let [resources (filter is-plugin-file? (g/node-value workspace :resource-list))]
+    resources))
+
+(defn unpack-resource! [workspace resource]
+  (let [target-path (plugin-path workspace (resource/proj-path resource))
+        parent-dir (.getParentFile ^File target-path)
+        input-stream (clojure.java.io/input-stream resource)]
+    (when-not (.exists parent-dir)
+      (.mkdirs parent-dir))
+    (io/copy input-stream target-path)))
+
+; It's important to use the same class loader, so that the type signatures match
+(def class-loader (clojure.lang.DynamicClassLoader. (.getContextClassLoader (Thread/currentThread))))
+
+(defn load-class! [class-name]
+  (Class/forName class-name true class-loader))
+
+(defn- register-jar-file! [workspace resource]
+  (let [jar-file (plugin-path workspace (resource/proj-path resource))]
+    (.addURL ^clojure.lang.DynamicClassLoader class-loader (io/as-url jar-file))))
+
+(defn unpack-editor-plugins! [workspace changed]
+  ; Used for unpacking the .jar files and shared libraries (.so, .dylib, .dll) to disc
+  ; TODO: Handle removed plugins (e.g. a dependency was removed)
+  (let [changed-resources (set (map resource/proj-path changed))
+        all-plugin-resources (find-plugins-shared-libraries workspace)
+        changed-plugin-resources (filter (fn [x] (contains? changed-resources (resource/proj-path x))) all-plugin-resources)
+        changed-jar-resources (filter is-jar-file? changed-plugin-resources)]
+    (dorun (map (fn [x] (unpack-resource! workspace x)) changed-plugin-resources))
+    (dorun (map (fn [x] (register-jar-file! workspace x)) changed-jar-resources))))
```

Comment:
These could also be written as `doseq` expressions to make them look more like an imperative `for`-loop:
```clj
(doseq [x changed-plugin-resources]
  (unpack-resource! workspace x))
```